File: canaries/python/sql_injection.py
import sqlite3
import mysql.connector
import psycopg2
from flask import Flask, request

app = Flask(__name__)

# SQLi vulnerability #1: Direct string concatenation
@app.route('/users')
def get_users():
    user_id = request.args.get('id')
    conn = sqlite3.connect('example.db')
    cursor = conn.cursor()

    # Vulnerable: Direct string concatenation
    query = "SELECT * FROM users WHERE id = " + user_id
    cursor.execute(query)
    results = cursor.fetchall()

    return str(results)

# SQLi vulnerability #2: String formatting
@app.route('/search')
def search_users():
    search_term = request.args.get('q')
    conn = sqlite3.connect('example.db')
    cursor = conn.cursor()

    # Vulnerable: String formatting
    query = "SELECT * FROM users WHERE name LIKE '%{}%'".format(search_term)
    cursor.execute(query)
    results = cursor.fetchall()

    return str(results)

# SQLi vulnerability #3: f-string injection
@app.route('/login')
def login():
    username = request.form.get('username')
    password = request.form.get('password')

    conn = mysql.connector.connect(
        host='localhost',
        user='root',
        password='password',
        database='testdb'
    )
    cursor = conn.cursor()

    # Vulnerable: f-string injection
    query = f"SELECT * FROM users WHERE username='{username}' AND password='{password}'"
    cursor.execute(query)
    result = cursor.fetchone()

    return str(result)

# SQLi vulnerability #4: % formatting
@app.route('/admin')
def admin_panel():
    role = request.args.get('role')
    conn = psycopg2.connect(
        host="localhost",
        database="testdb",
        user="admin",
        password="secret"
    )
    cursor = conn.cursor()

    # Vulnerable: % string formatting
    query = "SELECT * FROM admin_users WHERE role = '%s'" % role
    cursor.execute(query)
    results = cursor.fetchall()

    return str(results)

if __name__ == '__main__':
    app.run(debug=True)

File: canaries/golang/sql_injection.go
package main

import (
	"database/sql"
	"fmt"
	"net/http"

	_ "github.com/lib/pq"
)

var db *sql.DB

// SQL Injection vulnerability #1: Direct string concatenation
func getUserByID(w http.ResponseWriter, r *http.Request) {
	userID := r.URL.Query().Get("id")

	// Vulnerable: Direct string concatenation
	query := "SELECT username, email FROM users WHERE id = " + userID

	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, "Database error", 500)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var username, email string
		rows.Scan(&username, &email)
		fmt.Fprintf(w, "User: %s, Email: %s\n", username, email)
	}
}

// SQL Injection vulnerability #2: fmt.Sprintf
func searchUsers(w http.ResponseWriter, r *http.Request) {
	searchTerm := r.URL.Query().Get("search")

	// Vulnerable: Using fmt.Sprintf for query construction
	query := fmt.Sprintf("SELECT * FROM users WHERE username LIKE '%%%s%%'", searchTerm)

	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, "Database error", 500)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var id int
		var username, email string
		rows.Scan(&id, &username, &email)
		fmt.Fprintf(w, "ID: %d, User: %s, Email: %s\n", id, username, email)
	}
}

// SQL Injection vulnerability #3: Dynamic ORDER BY
func listUsers(w http.ResponseWriter, r *http.Request) {
	orderBy := r.URL.Query().Get("order")
	if orderBy == "" {
		orderBy = "username"
	}

	// Vulnerable: Dynamic ORDER BY without validation
	query := "SELECT id, username, email FROM users ORDER BY " + orderBy

	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, "Database error", 500)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var id int
		var username, email string
		rows.Scan(&id, &username, &email)
		fmt.Fprintf(w, "ID: %d, User: %s, Email: %s\n", id, username, email)
	}
}

// SQL Injection vulnerability #4: Authentication bypass
func login(w http.ResponseWriter, r *http.Request) {
	username := r.FormValue("username")
	password := r.FormValue("password")

	// Vulnerable: Direct concatenation in WHERE clause
	query := fmt.Sprintf("SELECT id FROM users WHERE username='%s' AND password='%s'", username, password)

	var userID int
	err := db.QueryRow(query).Scan(&userID)
	if err != nil {
		fmt.Fprintf(w, "Login failed")
		return
	}

	fmt.Fprintf(w, "Login successful! User ID: %d", userID)
}

// SQL Injection vulnerability #5: UNION-based injection
func getProductInfo(w http.ResponseWriter, r *http.Request) {
	productID := r.URL.Query().Get("product_id")

	// Vulnerable: Allows UNION attacks
	query := "SELECT name, price FROM products WHERE id = " + productID

	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, "Database error", 500)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var name, price string
		rows.Scan(&name, &price)
		fmt.Fprintf(w, "Product: %s, Price: %s\n", name, price)
	}
}

// SQL Injection vulnerability #6: Batch operations
func updateUserRoles(w http.ResponseWriter, r *http.Request) {
	userIDs := r.FormValue("user_ids") // e.g., "1,2,3"
	role := r.FormValue("role")

	// Vulnerable: IN clause with direct insertion
	query := fmt.Sprintf("UPDATE users SET role = '%s' WHERE id IN (%s)", role, userIDs)

	_, err := db.Exec(query)
	if err != nil {
		http.Error(w, "Database error", 500)
		return
	}

	fmt.Fprintf(w, "Users updated successfully")
}

func main() {
	// Initialize database connection (would be real in practice)
	// db, _ = sql.Open("postgres", "user=username dbname=mydb sslmode=disable")

	http.HandleFunc("/user", getUserByID)
	http.HandleFunc("/search", searchUsers)
	http.HandleFunc("/list", listUsers)
	http.HandleFunc("/login", login)
	http.HandleFunc("/product", getProductInfo)
	http.HandleFunc("/update-roles", updateUserRoles)

	fmt.Println("Starting vulnerable server on :8081")
	http.ListenAndServe(":8081", nil)
}